cloudinit/config/modules.py

```diff
@@ -35,28 +38,31 @@ def form_module_name(name):
     return canon_name
 
 
-def fixup_module(mod, def_freq=PER_INSTANCE):
-    if not hasattr(mod, "frequency"):
-        setattr(mod, "frequency", def_freq)
-    else:
-        freq = mod.frequency
-        if freq and freq not in FREQUENCIES:
-            LOG.warning("Module %s has an unknown frequency %s", mod, freq)
-    if not hasattr(mod, "distros"):
-        setattr(mod, "distros", [])
-    if not hasattr(mod, "osfamilies"):
-        setattr(mod, "osfamilies", [])
-    return mod
+def validate_module(mod, name):
+    if (
+        not hasattr(mod, "meta")
+        or "frequency" not in mod.meta
+        or "distros" not in mod.meta
+    ):
+        raise ValueError(
+            f"Module '{mod}' with name '{name}' MUST have a 'meta' attribute "
+            "of type 'MetaSchema'."
+        )
+    if mod.meta["frequency"] not in FREQUENCIES:
+        raise ValueError(
+            f"Module '{mod}' with name '{name}' has an invalid frequency "
+            f"{mod.meta['frequency']}."
+        )
```

Comment:
Is it worth adding the following?
```suggestion
        )
    if hasattr(mod, "schema"):
        raise ValueError(f"Module '{mod}' with name '{name}' has a JSON "schema" attribute defined. Please define schema in cloud-init-schema,json.")
```